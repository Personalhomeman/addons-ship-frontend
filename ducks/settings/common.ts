import { actionTypeCreator } from '@/utils';
export const $ = actionTypeCreator('SETTINGS');
