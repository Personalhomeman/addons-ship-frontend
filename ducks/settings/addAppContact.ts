import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { RootState } from '@/store';
import { AppContact } from '@/models/settings';
import { ShipAPIService } from '@/services/ship-api';

import { $ } from './common';

const _addAppContact = (
  appSlug: string,
  email: string,
  notificationPreferences?: AppContact['notificationPreferences']
) => async (dispatch: Dispatch, _getState: () => RootState, { shipApi }: { shipApi: ShipAPIService }) => {
  dispatch(addAppContact.next());

  try {
    const appContact = await shipApi.addAppContact(appSlug, email, notificationPreferences);

    dispatch(addAppContact.complete(appContact));
  } catch (error) {
    dispatch(addAppContact.error(error));
  }
};

const addAppContact = Object.assign(_addAppContact, {
  next: createAction($`ADD_CONTACT_NEXT`),
  complete: createAction($`ADD_CONTACT_COMPLETE`, resolve => (appContact: AppContact) => resolve(appContact)),
  error: createAction($`ADD_CONTACT_ERROR`, resolve => (error: Error) => resolve(error))
});

export default addAppContact;
