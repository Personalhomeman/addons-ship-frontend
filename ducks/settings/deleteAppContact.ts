import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { RootState } from '@/store';
import { AppContact } from '@/models/settings';
import { ShipAPIService } from '@/services/ship-api';

import { $ } from './common';
const _deleteAppContact = (appSlug: string, appContact: AppContact) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(deleteAppContact.next());

  try {
    await shipApi.deleteAppContact(appSlug, appContact);

    dispatch(deleteAppContact.complete(appContact.id));
  } catch (error) {
    dispatch(deleteAppContact.error(error));
  }
};

const deleteAppContact = Object.assign(_deleteAppContact, {
  next: createAction($`DELETE_APP_CONTACT_NEXT`),
  complete: createAction($`DELETE_APP_CONTACT_COMPLETE`, resolve => (appContactId: string) => resolve(appContactId)),
  error: createAction($`DELETE_APP_CONTACT_ERROR`, resolve => (error: Error) => resolve(error))
});

export default deleteAppContact;
