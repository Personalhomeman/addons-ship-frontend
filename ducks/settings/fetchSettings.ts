import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { RootState } from '@/store';
import { Settings } from '@/models/settings';
import { ShipAPIService } from '@/services/ship-api';

import { $ } from './common';

const _fetchSettings = (appSlug: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(fetchSettings.next());

  try {
    const settings = await shipApi.getSettings(appSlug);

    dispatch(fetchSettings.complete(settings));
  } catch (error) {
    dispatch(fetchSettings.error(error));
  }
};

const fetchSettings = Object.assign(_fetchSettings, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (settings: Settings) => resolve(settings)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

export default fetchSettings;
