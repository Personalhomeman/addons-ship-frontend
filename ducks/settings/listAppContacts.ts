import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { RootState } from '@/store';
import { AppContact } from '@/models/settings';
import { ShipAPIService } from '@/services/ship-api';

import { $ } from './common';

const _listAppContacts = (appSlug: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(listAppContacts.next());

  try {
    const appContacts = await shipApi.listAppContacts(appSlug);

    dispatch(listAppContacts.complete(appContacts));
  } catch (error) {
    dispatch(listAppContacts.error(error));
  }
};

const listAppContacts = Object.assign(_listAppContacts, {
  next: createAction($`LIST_APP_CONTACTS_NEXT`),
  complete: createAction($`LIST_APP_CONTACTS_COMPLETE`, resolve => (appContacts: AppContact[]) => resolve(appContacts)),
  error: createAction($`LIST_APP_CONTACTS_ERROR`, resolve => (error: Error) => resolve(error))
});

export default listAppContacts;
