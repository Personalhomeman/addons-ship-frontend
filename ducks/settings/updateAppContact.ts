import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { RootState } from '@/store';
import { AppContact } from '@/models/settings';
import { ShipAPIService } from '@/services/ship-api';

import { $ } from './common';
const _updateAppContact = (appSlug: string, appContact: AppContact) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(updateAppContact.next());

  try {
    const newAppContact = await shipApi.updateAppContactNotificationPreferences(appSlug, appContact);

    dispatch(updateAppContact.complete(newAppContact));
  } catch (error) {
    dispatch(updateAppContact.error(error));
  }
};

const updateAppContact = Object.assign(_updateAppContact, {
  next: createAction($`UPDATE_APP_CONTACT_NEXT`),
  complete: createAction($`UPDATE_APP_CONTACT_COMPLETE`, resolve => (appContact: AppContact) => resolve(appContact)),
  error: createAction($`UPDATE_APP_CONTACT_ERROR`, resolve => (error: Error) => resolve(error))
});

export default updateAppContact;
