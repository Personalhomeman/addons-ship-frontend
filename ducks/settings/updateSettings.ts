import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { RootState } from '@/store';
import { Settings } from '@/models/settings';
import { ShipAPIService } from '@/services/ship-api';

import { $ } from './common';
const _updateSettings = (settings: Settings) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(updateSettings.next());

  try {
    await shipApi.updateSettings(settings);

    dispatch(updateSettings.complete(settings));
  } catch (error) {
    dispatch(updateSettings.error(error));
  }
};

const updateSettings = Object.assign(_updateSettings, {
  next: createAction($`UPDATE_NEXT`),
  complete: createAction($`UPDATE_COMPLETE`, resolve => (settings: Settings) => resolve(settings)),
  error: createAction($`UPDATE_ERROR`, resolve => (error: Error) => resolve(error))
});

export default updateSettings;
