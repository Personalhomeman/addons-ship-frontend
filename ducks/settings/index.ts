import { createReducer } from 'deox';

import { Settings, AppContact } from '@/models/settings';

import fetchSettings from './fetchSettings';
import updateSettings from './updateSettings';
import addAppContact from './addAppContact';
import listAppContacts from './listAppContacts';

const defaultState: SettingsState = {
  settings: {
    projectType: 'other'
  },
  appContacts: []
};

export { fetchSettings, updateSettings, addAppContact, listAppContacts };

export type SettingsState = { settings: Settings; appContacts: AppContact[] };
export default createReducer(defaultState, handleAction => [
  handleAction(fetchSettings.complete, (state, { payload }) => ({ ...state, settings: payload })),
  handleAction(addAppContact.complete, (state, { payload }) => ({
    ...state,
    appContacts: state.appContacts.concat(payload)
  })),
  handleAction(listAppContacts.complete, (state, { payload }) => ({
    ...state,
    appContacts: payload
  }))
]);
