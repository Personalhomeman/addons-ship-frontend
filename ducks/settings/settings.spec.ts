jest.mock('@/services/ship-api');
jest.mock('@/utils/file');

import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store';
import thunk from 'redux-thunk';

import { mockSettings } from '@/mocks';
import { Settings, IosSettings, AppContact } from '@/models/settings';
import shipApi from '@/services/ship-api';

import settings, { fetchSettings, updateSettings, addAppContact } from '.';

describe('settings', () => {
  const mockAppContact: AppContact = {
    email: 'what@ev.er',
    isConfirmed: false,
    notificationPreferences: { newVersion: true, failedPublish: false, successfulPublish: true }
  };

  let mockStore: MockStoreCreator, store: MockStoreEnhanced;
  beforeEach(() => {
    mockStore = configureMockStore([thunk.withExtraArgument({ shipApi })]);
    store = mockStore();
  });

  describe('reducer', () => {
    it('loads settings', () => {
      const state = settings(undefined, fetchSettings.complete(mockSettings));

      expect(state).toMatchSnapshot();
    });

    it('adds an app contact', () => {
      const state = settings(
        {
          settings: { projectType: 'other' },
          appContacts: [mockAppContact]
        },
        addAppContact.complete({ email: 'bit@bot' } as AppContact)
      );

      expect(state).toMatchSnapshot();
    });
  });

  describe('fetchSettings', () => {
    it('fetches settings', async () => {
      const appSlug = 'app-slug';
      await store.dispatch(fetchSettings(appSlug) as any);

      expect(store.getActions()).toMatchSnapshot();
      expect(shipApi.getSettings).toHaveBeenCalledWith(appSlug);
    });

    it("can't fetch settings", async () => {
      (shipApi.getSettings as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(fetchSettings('app-slug') as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('updateSettings', () => {
    const updatedSettings: Settings = {
      ...mockSettings,
      iosSettings: { ...(mockSettings.iosSettings as IosSettings), artifactExposingWorkflows: 'Primary' }
    };

    it('updates settings', async () => {
      await store.dispatch(updateSettings(updatedSettings) as any);

      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't update settings", async () => {
      (shipApi.updateSettings as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(updateSettings(updatedSettings) as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('addAppContact', () => {
    it('adds a new contact', async () => {
      const appSlug = 'app-slug',
        email = 'some@random.email';
      await store.dispatch(addAppContact(appSlug, email) as any);

      expect(store.getActions()).toMatchSnapshot();
      expect(shipApi.addAppContact).toHaveBeenCalledWith(appSlug, email, undefined);
    });

    it('fails to add a new contact', async () => {
      (shipApi.addAppContact as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(addAppContact('app-clug', 'an@email.adr') as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
