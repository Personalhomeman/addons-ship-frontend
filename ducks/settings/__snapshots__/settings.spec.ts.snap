// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings addAppContact adds a new contact 1`] = `
Array [
  Object {
    "type": "SETTINGS_ADD_CONTACT_NEXT",
  },
  Object {
    "type": "SETTINGS_ADD_CONTACT_COMPLETE",
  },
]
`;

exports[`settings addAppContact fails to add a new contact 1`] = `
Array [
  Object {
    "type": "SETTINGS_ADD_CONTACT_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "SETTINGS_ADD_CONTACT_ERROR",
  },
]
`;

exports[`settings fetchSettings can't fetch settings 1`] = `
Array [
  Object {
    "type": "SETTINGS_GET_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "SETTINGS_GET_ERROR",
  },
]
`;

exports[`settings fetchSettings fetches settings 1`] = `
Array [
  Object {
    "type": "SETTINGS_GET_NEXT",
  },
  Object {
    "type": "SETTINGS_GET_COMPLETE",
  },
]
`;

exports[`settings listAppContacts fails to list app contacts 1`] = `
Array [
  Object {
    "type": "SETTINGS_LIST_APP_CONTACTS_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "SETTINGS_LIST_APP_CONTACTS_ERROR",
  },
]
`;

exports[`settings listAppContacts lists app contacts 1`] = `
Array [
  Object {
    "type": "SETTINGS_LIST_APP_CONTACTS_NEXT",
  },
  Object {
    "type": "SETTINGS_LIST_APP_CONTACTS_COMPLETE",
  },
]
`;

exports[`settings reducer adds an app contact 1`] = `
Object {
  "appContacts": Array [
    Object {
      "email": "what@ev.er",
      "isConfirmed": false,
      "notificationPreferences": Object {
        "failedPublish": false,
        "newVersion": true,
        "successfulPublish": true,
      },
    },
    Object {
      "email": "bit@bot",
    },
  ],
  "settings": Object {
    "projectType": "other",
  },
}
`;

exports[`settings reducer loads app contacts 1`] = `
Object {
  "appContacts": Array [
    Object {
      "email": "purr@req.io",
    },
    Object {
      "email": "bit@bot.io",
    },
  ],
  "settings": Object {
    "projectType": "other",
  },
}
`;

exports[`settings reducer loads settings 1`] = `
Object {
  "appContacts": Array [],
  "settings": Object {
    "androidSettings": Object {
      "artifactExposingWorkflows": "All",
      "selectedKeystoreFile": Object {
        "name": "abcdefghijkl1234567",
      },
      "selectedServiceAccountJsonFile": Object {
        "name": "abcdefghijkl12345671566",
      },
      "track": "Release",
    },
    "certificates": Array [
      Object {
        "name": "iPhone Developer: John Doe (ABCD1234)",
      },
      Object {
        "name": "iPhone Developer: John Doe (ABCD5678)",
      },
    ],
    "iosSettings": Object {
      "appSku": "Fill",
      "appSpecificPassword": "Fill",
      "appleDeveloperAccountEmail": "Fill",
      "artifactExposingWorkflows": "All",
      "selectedCertificate": Object {
        "name": "iPhone Developer: John Doe (ABCD5678)",
      },
      "selectedProvProfile": Object {
        "name": "abcdefghijkl1234567",
      },
    },
    "keystoreFiles": Array [
      Object {
        "name": "abcdefghijkl1234567",
      },
      Object {
        "name": "abcdefghijkl12345671566",
      },
    ],
    "projectType": "other",
    "provProfiles": Array [
      Object {
        "name": "abcdefghijkl1234567",
      },
      Object {
        "name": "abcdefghijkl12345671566",
      },
    ],
    "serviceAccountJsonFiles": Array [
      Object {
        "name": "abcdefghijkl1234567",
      },
      Object {
        "name": "abcdefghijkl12345671566",
      },
    ],
  },
}
`;

exports[`settings updateSettings can't update settings 1`] = `
Array [
  Object {
    "type": "SETTINGS_UPDATE_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "SETTINGS_UPDATE_ERROR",
  },
]
`;

exports[`settings updateSettings updates settings 1`] = `
Array [
  Object {
    "type": "SETTINGS_UPDATE_NEXT",
  },
  Object {
    "payload": Object {
      "androidSettings": Object {
        "artifactExposingWorkflows": "All",
        "selectedKeystoreFile": Object {
          "name": "abcdefghijkl1234567",
        },
        "selectedServiceAccountJsonFile": Object {
          "name": "abcdefghijkl12345671566",
        },
        "track": "Release",
      },
      "certificates": Array [
        Object {
          "name": "iPhone Developer: John Doe (ABCD1234)",
        },
        Object {
          "name": "iPhone Developer: John Doe (ABCD5678)",
        },
      ],
      "iosSettings": Object {
        "appSku": "Fill",
        "appSpecificPassword": "Fill",
        "appleDeveloperAccountEmail": "Fill",
        "artifactExposingWorkflows": "Primary",
        "selectedCertificate": Object {
          "name": "iPhone Developer: John Doe (ABCD5678)",
        },
        "selectedProvProfile": Object {
          "name": "abcdefghijkl1234567",
        },
      },
      "keystoreFiles": Array [
        Object {
          "name": "abcdefghijkl1234567",
        },
        Object {
          "name": "abcdefghijkl12345671566",
        },
      ],
      "projectType": "other",
      "provProfiles": Array [
        Object {
          "name": "abcdefghijkl1234567",
        },
        Object {
          "name": "abcdefghijkl12345671566",
        },
      ],
      "serviceAccountJsonFiles": Array [
        Object {
          "name": "abcdefghijkl1234567",
        },
        Object {
          "name": "abcdefghijkl12345671566",
        },
      ],
    },
    "type": "SETTINGS_UPDATE_COMPLETE",
  },
]
`;
